spectral_cube/io/casa_image.py

```diff
@@ -356,6 +255,173 @@ def load_casa_image(filename, skipdata=False,
     return normalize_cube_stokes(cube, target_cls=target_cls)
 
 
+class MemmapWrapper:
+    """
+    A wrapper class for dask that opens memmap only when __getitem__ is called,
+    which prevents issues with too many open files if opening a memmap for all
+    chunks at the same time.
+    """
+
+    def __init__(self, filename, **kwargs):
+        self._filename = filename
+        self._kwargs = kwargs
+        self.shape = kwargs['shape'][::-1]
+        self.dtype = kwargs['dtype']
+        self.ndim = len(self.shape)
+
+    def __getitem__(self, item):
+        # We open a file manually and return an in-memory copy of the array
+        # otherwise the file doesn't get closed properly.
+        with open(self._filename) as f:
+            return np.memmap(f, mode='readonly', order='F', **self._kwargs).T[item].copy()
```

Comment:
This is kind of a mess, but I see why it's needed.

There must be some way to define a `np.memmap`-like wrapper around `mmap.mmap`, the lower-level python function, that internally translates the slice syntax to the appropriate locations.  I'm not sure whether that would be more performant, but it would at least prevent the many-open-files problems.

Reply:
Yes that could be a solution, I'll investigate whether it helpers with performance.